From fa3d5e5b9046d62c053012d6d2ea4165e0197f95 Mon Sep 17 00:00:00 2001
From: Eberhard Beilharz <eb1@sil.org>
Date: Fri, 30 Mar 2012 18:14:33 +0200
Subject: [PATCH] Xamarin-4158: Fix off-screen calculation of
 ToolStripDropDown

In a multi-monitor situation the ToolStripDropDown has to use the
dimensions of the monitor the position is located on rather than
the virtual screen dimensions.

This means that the drop down is always limited to one monitor, but
that is prefered over not displaying some items because they are
off-screen if the displaying monitor is smaller.

Change-Id: I18bf04e5a97be7be91af3366cfcf813abbba0eff
---
 .../System.Windows.Forms/ToolStripDropDown.cs      |   32 +++++++++++--------
 1 files changed, 18 insertions(+), 14 deletions(-)

Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStripDropDown.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStripDropDown.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStripDropDown.cs
@@ -399,14 +399,18 @@
 		{
 			Show (new Point (x, y), DefaultDropDownDirection);
 		}
-		
+
+		// I don't think primary to the right of secondary (which gives us negative position
+		// for a point on the secondary monitor) can happen on Linux since it seems that
+		// primary is always the left-most monitor - at least with my testing in Ubuntu 11.10.
+		[MonoTODO ("Need to deal with multi-monitor situation where primary is to the right of secondary monitor")]
 		public void Show (Point position, ToolStripDropDownDirection direction)
 		{
 			this.PerformLayout ();
 			
 			Point show_point = position;
-			Point max_screen = new Point (SystemInformation.VirtualScreen.Width, SystemInformation.VirtualScreen.Height);
-			
+			Rectangle max_screen = Screen.GetWorkingArea (position);
+
 			if (this is ContextMenuStrip) {
 				// If we are going to go offscreen, adjust our direction so we don't...
 				// X direction
@@ -424,16 +428,16 @@
 							direction = ToolStripDropDownDirection.Right;
 						break;
 					case ToolStripDropDownDirection.AboveRight:
-						if (show_point.X + this.Width > max_screen.X)
+						if (show_point.X + this.Width > max_screen.Right)
 							direction = ToolStripDropDownDirection.AboveLeft;
 						break;
 					case ToolStripDropDownDirection.BelowRight:
 					case ToolStripDropDownDirection.Default:
-						if (show_point.X + this.Width > max_screen.X)
+						if (show_point.X + this.Width > max_screen.Right)
 							direction = ToolStripDropDownDirection.BelowLeft;
 						break;
 					case ToolStripDropDownDirection.Right:
-						if (show_point.X + this.Width > max_screen.X)
+						if (show_point.X + this.Width > max_screen.Right)
 							direction = ToolStripDropDownDirection.Left;
 						break;
 				}
@@ -449,20 +453,20 @@
 							direction = ToolStripDropDownDirection.BelowRight;
 						break;
 					case ToolStripDropDownDirection.BelowLeft:
-						if (show_point.Y + this.Height > max_screen.Y && show_point.Y - this.Height > 0)
+						if (show_point.Y + this.Height > max_screen.Bottom && show_point.Y - this.Height > 0)
 							direction = ToolStripDropDownDirection.AboveLeft;
 						break;
 					case ToolStripDropDownDirection.BelowRight:
 					case ToolStripDropDownDirection.Default:
-						if (show_point.Y + this.Height > max_screen.Y && show_point.Y - this.Height > 0)
+						if (show_point.Y + this.Height > max_screen.Bottom && show_point.Y - this.Height > 0)
 							direction = ToolStripDropDownDirection.AboveRight;
 						break;
 					case ToolStripDropDownDirection.Left:
-						if (show_point.Y + this.Height > max_screen.Y && show_point.Y - this.Height > 0)
+						if (show_point.Y + this.Height > max_screen.Bottom && show_point.Y - this.Height > 0)
 							direction = ToolStripDropDownDirection.AboveLeft;
 						break;
 					case ToolStripDropDownDirection.Right:
-						if (show_point.Y + this.Height > max_screen.Y && show_point.Y - this.Height > 0)
+						if (show_point.Y + this.Height > max_screen.Bottom && show_point.Y - this.Height > 0)
 							direction = ToolStripDropDownDirection.AboveRight;
 						break;
 				}
@@ -487,14 +491,14 @@
 			}
 
 			// Fix offscreen horizontal positions
-			if ((show_point.X + this.Width) > max_screen.X)
-				show_point.X = max_screen.X - this.Width;
+			if ((show_point.X + this.Width) > max_screen.Right)
+				show_point.X = max_screen.Right - this.Width;
 			if (show_point.X < 0)
 				show_point.X = 0;
 
 			// Fix offscreen vertical positions
-			if ((show_point.Y + this.Height) > max_screen.Y)
-				show_point.Y = max_screen.Y - this.Height;
+			if ((show_point.Y + this.Height) > max_screen.Bottom)
+				show_point.Y = max_screen.Bottom - this.Height;
 			if (show_point.Y < 0)
 				show_point.Y = 0;
 
